File: src/com/douma/_13_day/_170/_170_TwoSum1.py
class TwoSum:

    def __init__(self):
        self.nums = []
        self.is_sorted = False

    def add(self, number: int) -> None:
        self.nums.append(number)
        self.is_sorted = False


    def find(self, value: int) -> bool:
        if not self.is_sorted:
            self.nums.sort()
            self.is_sorted = True

        left, right = 0, len(self.nums) - 1
        while left < right:
            lr_sum = self.nums[left] + self.nums[right]
            if lr_sum == value:
                return True
            elif lr_sum < value:
                left += 1
            else:
                right -= 1
        return False


# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum()
# obj.add(number)
# param_2 = obj.find(value)

File: src/com/douma/_13_day/_170/_170_TwoSum1.go
type TwoSum struct {
    nums []int
    isSorted bool
}


func Constructor() TwoSum {
    nums := make([]int, 0)
    return TwoSum{nums:nums, isSorted:false}
}


func (this *TwoSum) Add(number int)  {
    this.nums = append(this.nums, number)
    this.isSorted = false
}


func (this *TwoSum) Find(value int) bool {
    if !this.isSorted {
        sort.Ints(this.nums)
        this.isSorted = true
    }

    var left, right = 0, len(this.nums) - 1
    for left < right {
        var sum = this.nums[left] + this.nums[right]
        if sum == value {
            return true
        } else if sum < value {
            left++
        } else {
            right--
        }
    }

    return false
}


/**
 * Your TwoSum object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Add(number);
 * param_2 := obj.Find(value);
 */